website/registrations/utils.py

```diff
@@ -0,0 +1,445 @@
+                        'required': schema_block.get('required'),
+                    }
+                })
+            elif schema_block['block_type'] in ('short-text-input', 'long-text-input'):
+                validations.update({
+                    schema_block['registration_response_key']: {
+                        'type': 'string',
+                        'required': schema_block.get('required'),
+                    }
+                })
+        return validations
+
+    def log_error(self, **kwargs):
+        self.errors.append({
+            'header': kwargs['header'],
+            'column_index': get_excel_column_name(kwargs['column_index']),
+            'row_index': kwargs['row_index'],
+            'external_id': kwargs.get('external_id', ''),
+            'type': kwargs.get('type', '')
+        })
+
+    def validate_csv_header_list(self):
+        expected_headers = self.validations.keys()
+        actual_headers = self.headers
+        invalid_headers = list(set(actual_headers) - set(expected_headers))
+        missing_headers = list(set(expected_headers) - set(actual_headers))
+        if invalid_headers or missing_headers:
+            raise InvalidHeadersError({'invalid_headers': invalid_headers, 'missing_headers': missing_headers})
+
+    def get_parsed(self):
+        parsed = []
+        for row in self.rows:
+            parsed.append({'csv_raw': row.get_raw_value(), 'csv_parsed': row.get_parsed_value()})
+        return {'schema_id': self.schema_id, 'registrations': parsed}
+
+    def validate(self):
+        for row in self.rows:
+            row.validate()
+
+class Row():
+    @property
+    def is_validated(self):
+        return all([cell.is_validated for cell in self.cells])
+
+    def __init__(self, row_dict, validations, store, log_error):
+        self.row_dict = row_dict
+        self.cells = [Cell(header,
+                           value, validations[header], store,
+                           functools.partial(log_error, external_id=row_dict.get('External ID', ''), column_index=column_index))
+                           for column_index, (header, value) in enumerate(row_dict.items())]
+
+    def get_metadata(self):
+        parsed_metadata = {}
+        for cell in self.cells:
+            if cell.is_metadata:
+                parsed_metadata.update(cell.get_parsed_value())
+        return parsed_metadata
+
+    def get_registration_responses(self):
+        parsed_responses = {}
+        for cell in self.cells:
+            if not cell.is_metadata:
+                parsed_responses.update(cell.get_parsed_value())
+        return parsed_responses
+
+    def get_parsed_value(self):
+        return {'metadata': self.get_metadata(),
+                'registration_responses': self.get_registration_responses()}
+
+    def get_raw_value(self):
+        raw_value_buffer = io.StringIO()
+        csv_writer = csv.writer(raw_value_buffer)
+        cell_values = [cell.value for cell in self.cells]
```

Comment:
The csv writer has only a `writerow` function, which takes a list of cells in a row as parameter. There is no available method  to write just a cell. That means if we implement a `get_raw_value()` method for `Cell`, it would not really be returning the raw value of the cell here, unless we manually reconstruct, without using a csv writer, the raw value for the cell, in which case, we wouldn't be using the `writerow` method of the csv writer at all.

Reply:
Cell already defines `get_raw_value` (see L235), it returns `cell.value`. To me (and maybe only me 😅 ), it makes more sense for this to be `cell_values = [cell.get_raw_value() for cell in self.cells]` so that cell.raw_value and cell.value have room to change in the future -- i.e., cell.value becomes a property that surfaces a sanitized or otherwise formatted version of the raw value, while raw_value does not.

Very possibly a premature optimization that I'm primarily advocating on the basis of semantics. Feel free to ignore.